api/src/test/java/org/openmrs/scheduler/tasks/AutoRetireUsersTaskTest.java

```diff
@@ -0,0 +1,130 @@
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class AutoRetireUsersTaskTest extends BaseContextSensitiveTest {
+	protected static final String XML_FILENAME = "org/openmrs/api/include/UserServiceTest.xml";
+	private static final long ONE_DAY_IN_MILLISECONDS = 24 * 60 * 60 * 1000;
+
+	private UserService userService;
+	private AdministrationService administrationService;
+	private AutoRetireUsersTask autoRetireUsersTask;
+
+	@BeforeEach
+	public void setup() {
+		userService = Context.getUserService();
+		administrationService = Context.getAdministrationService();
+		autoRetireUsersTask = new AutoRetireUsersTask();
+		executeDataSet(XML_FILENAME);
+	}
+
+	@Test
+	public void getUsersToRetire_shouldReturnUsersThatExceedInactivityThreshold() {
+		// Global property for number of days to retire
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, "1")); // 1 day
+
+		User inactiveUser = userService.getUser(1);
+		inactiveUser.setUserProperty(
+			OpenmrsConstants.USER_PROPERTY_LAST_LOGIN_TIMESTAMP, 
+			String.valueOf(System.currentTimeMillis() - (2 * ONE_DAY_IN_MILLISECONDS))
+		); // 2 days ago
+
+		userService.saveUser(inactiveUser);
+
+		Set<User> usersToRetire = autoRetireUsersTask.getUsersToRetire(userService);
+
+		assertThat(usersToRetire, contains(inactiveUser));
+	}
+
+	@Test
+	public void getUsersToRetire_shouldNotReturnUsersThatDoNotExceedInactivityThreshold() {
+		// Global property for number of days to retire
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, "2")); // 1 day
+
+		User inactiveUser = userService.getUser(1);
+		inactiveUser.setUserProperty(
+			OpenmrsConstants.USER_PROPERTY_LAST_LOGIN_TIMESTAMP, 
+			String.valueOf(System.currentTimeMillis() - (ONE_DAY_IN_MILLISECONDS))
+		); // 1 day ago
+
+		userService.saveUser(inactiveUser);
+
+		Set<User> usersToRetire = autoRetireUsersTask.getUsersToRetire(userService);
+
+		assertThat(usersToRetire, not(contains(inactiveUser)));
+	}
+
+	@Test
+	public void getUsersToRetire_shouldNotReturnAlreadyRetiredUsers() {
+		// Global property for number of days to retire
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, "1")); // 1 day
+
+		User retiredUser = userService.getUser(1);
+		retiredUser.setUserProperty(
+			OpenmrsConstants.USER_PROPERTY_LAST_LOGIN_TIMESTAMP, 
+			String.valueOf(System.currentTimeMillis() - (2 * ONE_DAY_IN_MILLISECONDS))
+		); // 2 days ago
+		
+		retiredUser.setRetired(true);
+		retiredUser.setRetireReason("User retired due to inactivity");
+
+		userService.saveUser(retiredUser);
+
+		Set<User> usersToRetire = autoRetireUsersTask.getUsersToRetire(userService);
+
+		assertThat(usersToRetire, not(contains(retiredUser)));
```

Comment:
Is the list of usersToRetire expected to be non empty?

Reply:
If another is added (in XML_FILENAME) then this test will fail because the list is not empty.

Reply:
Your reasoning makes a lot of sense! 👍